runtime/Java/src/org/antlr/v4/runtime/tree/Trees.java

```diff
@@ -97,6 +132,36 @@ else if ( t instanceof TerminalNode) {
 		return t.getPayload().toString();
 	}
 
+	/** @since 4.10.2 */
+	public static String getJSONNodeText(Tree t, List<String> ruleNames) {
+		if ( ruleNames!=null ) {
+			if ( t instanceof RuleContext ) {
+				int ruleIndex = ((RuleContext)t).getRuleContext().getRuleIndex();
+				String ruleName = ruleNames.get(ruleIndex);
+				int altNumber = ((RuleContext) t).getAltNumber();
+				if ( altNumber!=ATN.INVALID_ALT_NUMBER ) {
+					return ruleName+":"+altNumber;
+				}
+				return '"'+ruleName+'"';
+			}
+			else if ( t instanceof ErrorNode) {
+				return t.toString();
+			}
+			else if ( t instanceof TerminalNode) {
+				Token symbol = ((TerminalNode)t).getSymbol();
+				if (symbol != null) {
+					return String.format("{\"idx\":\"%d\",\"text\":\"%s\"}", symbol.getTokenIndex(), symbol.getText());
```

Comment:
Text can be restored from token that accesses by index. Is it necessary info?

Reply:
Was wondering about that. Means we need another data structure, `tokens`, to be sent as well. Might not be bad idea.

Reply:
Depends if we want a serialization or just a tree for display I think. both?

Reply:
if we use just idx then it's just list of ints to store terminal nodes as children of a subtree root :)

Reply:
> Was wondering about that. Means we need another data structure, `tokens`, to be sent as well. Might not be bad idea.

I'd say `tree`, `tokens` and `input`. Token may contain offset and length instead of text.

Reply:
ultimately yep, we need a big blob of stuff for server...might as well bundle up as you say. We need rulenames too.

Reply:
I would definitely recommend considering what you'd want to do with the tree. For Trash, I have to do a full serialization of not only the parse tree, but the tokstream + charstream + parser + lexer. I have to have everything because each tool can modify the tree. For example, if I delete a node in the parse tree, I'll need to adjust the tokstream and charstream in order to have a consistent set of data structures from one pipelined app to the next. For a proper "xpath" query system on the tree, one may want to specify or extract the column/line numbers in the query. I do have a separate tool to output as JSON or XML a simplified parse tree for display only, outputting only children for internal nodes and Text for the leaves. It comes in handy for [questions like this](https://stackoverflow.com/q/72838362/4779853). But, it's not intended to reconstruct the original text because there's no tokstream for off-default-channel content (e.g., comments or whitespace). If you offer something in the runtime, people need to be clear what it can be used for.

Reply:
yeah, i'm dumping it all now:

```
{
  "rules": [
    "s",
    "expr"
  ],
  "input": "99",
  "tokens": [
    {
      "type": 3,
      "line": 1,
      "pos": 0,
      "channel": 0,
      "text": "99"
    },
    {
      "type": -1,
      "line": 1,
      "pos": 2,
      "channel": 0,
      "text": "<EOF>"
    }
  ],
  "tree": {
    "0": [
      {
        "1": [
          0
        ]
      },
      1
    ]
  }
}
```
Still just an experiment...

Reply:
I'd make it an array of tree nodes, rather than just the root of the tree. If an array, it can represent results from an XPath search, which is a list of tree nodes.

Reply:
Hmm... Where are the point of this is to create a serialization form for a structured tree. Somehow those list of nodes would have to be linked up or are you simply suggesting that we also include a list of the nodes?